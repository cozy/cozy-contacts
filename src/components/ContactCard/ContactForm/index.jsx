import React from 'react'
import PropTypes from 'prop-types'
import { Form } from 'react-final-form'
import arrayMutators from 'final-form-arrays'
import { translate } from 'cozy-ui/transpiled/react/I18n'

import FieldsetTitle from '../../Common/FieldsetTitle'
import ContactFormField from './ContactFormField'
import ContactFieldInput from '../ContactFieldInput'
import { fullContactPropTypes } from '../../ContactPropTypes'
import contactToFormValues from './contactToFormValues'
import formValuesToContact from './formValuesToContact'
import { fields } from './fieldsConfig'

// this variable will be set in the form's render prop
// and used by the submit button in ContactFormModal
// to be able to trigger the submit from outside the form
// See react-final-form examples here: https://www.npmjs.com/package/react-final-form#external-submit
let _submitContactForm

function setSubmitContactForm(handleSubmit) {
  _submitContactForm = handleSubmit
}

export function getSubmitContactForm() {
  return _submitContactForm
}

const ContactForm = ({ contact, onSubmit, t }) => (
  <Form
    mutators={{ ...arrayMutators }}
    onSubmit={data => onSubmit(formValuesToContact(data, contact))}
    initialValues={contactToFormValues(contact, t)}
    render={({ handleSubmit }) => {
      setSubmitContactForm(handleSubmit)
      return (
        <div>
          <form onSubmit={handleSubmit} className="u-flex u-flex-column">
            <FieldsetTitle title={t('contact_info')} />
            {fields.map(
              ({
                name,
                icon,
                type,
                required,
                hasLabel,
                isArray,
                labelProps,
                isMultiline
              }) => (
                <ContactFormField
                  key={name}
                  name={name}
                  icon={icon}
                  label={t(`fields.${name}`)}
                  t={t}
                  isArray={isArray}
                  renderInput={(inputName, id) => (
                    <ContactFieldInput
                      id={id}
                      name={inputName}
                      type={type}
                      label={t(`fields.${name}`)}
                      required={required}
                      withLabel={hasLabel}
                      labelPlaceholder={t('fields.label')}
                      labelProps={labelProps}
                      isMultiline={isMultiline}
                    />
                  )}
                />
              )
            )}
          </form>
        </div>
      )
    }}
  />
)

ContactForm.propTypes = {
  contact: fullContactPropTypes,
  onSubmit: PropTypes.func.isRequired,
  t: PropTypes.func.isRequired
}

export default translate()(ContactForm)
