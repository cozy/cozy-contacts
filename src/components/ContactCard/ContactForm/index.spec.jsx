import React from 'react'
import { mount } from 'enzyme'
import ContactForm from './index'
import { I18n } from 'cozy-ui/transpiled/react/I18n'

import { render, fireEvent } from '@testing-library/react'
import { createMockClient } from 'cozy-client'
import AppLike from '../../../tests/Applike'

import langEn from '../../../locales/en.json'
import { johnDoeContact as contact } from '../../../helpers/testData'

const dictRequire = () => langEn

const client = createMockClient({})
const setup = ({ contact }) => {
  return {
    contact,
    onSubmit: () => jest.fn(),
    onCancel: () => jest.fn()
  }
}

const labels = [
  'Firstname',
  'Lastname',
  'Phone',
  'Email',
  'Address',
  'Cozy URL',
  'Company',
  'Job title',
  'Birthday',
  'Notes'
]

describe('ContactForm', () => {
  it(`should contains ${labels.length} fields and labels`, () => {
    const props = setup({ contact: {} })
    const jsx = (
      <AppLike client={client}>
        <ContactForm {...props} />
      </AppLike>
    )
    const { getByLabelText } = render(jsx)
    labels.map(label => {
      expect(getByLabelText(label)).toBeTruthy()
    })
  })

  it('should fill form with contact data', () => {
    const props = setup({ contact })
    const jsx = (
      <AppLike client={client}>
        <ContactForm {...props} />
      </AppLike>
    )
    const { getByDisplayValue } = render(jsx)
    expect(getByDisplayValue(contact.name.givenName)).toBeTruthy()
    expect(getByDisplayValue(contact.name.familyName)).toBeTruthy()
    expect(getByDisplayValue(contact.phone[0].number)).toBeTruthy()
    expect(getByDisplayValue(contact.email[0].address)).toBeTruthy()
    expect(getByDisplayValue(contact.address[0].formattedAddress)).toBeTruthy()
    expect(getByDisplayValue(contact.cozy[0].url)).toBeTruthy()
    expect(getByDisplayValue(contact.company)).toBeTruthy()
    expect(getByDisplayValue(contact.jobTitle)).toBeTruthy()
    expect(getByDisplayValue(contact.note)).toBeTruthy()
  })

  it('should submit a well formatted contact', () => {
    const expected = {
      address: [
        {
          formattedAddress: '18 rue des fleurs, Pecado',
          primary: true,
          type: undefined
        }
      ],
      birthday: '31/12/2015',
      company: 'Cozy CLoud',
      cozy: [
        { label: undefined, primary: true, url: 'https://jcvd.cozy.cloud' }
      ],
      displayName: 'Jean-Claude Van Cozy',
      email: [{ address: 'jcvc@cozy.cloud', primary: true, type: undefined }],
      fullname: 'Jean-Claude Van Cozy',
      indexes: {
        byFamilyNameGivenNameEmailCozyUrl:
          'Van CozyJean-Claudejcvc@cozy.cloudjcvd.cozy.cloud'
      },
      jobTitle: 'Dreamer',
      metadata: { cozy: true, version: 1 },
      name: { familyName: 'Van Cozy', givenName: 'Jean-Claude' },
      note: 'Whatever.',
      phone: [{ number: '+33678987654', primary: true, type: undefined }],
      relationships: { groups: { data: [] } }
    }
    const onSubmit = contact => {
      expect(contact).toEqual(expected)
    }

    const form = mount(
      <I18n lang="en" dictRequire={dictRequire}>
        <ContactForm onSubmit={onSubmit} onCancel={() => {}} />
      </I18n>
    )

    const fields = {
      givenName: 'Jean-Claude',
      familyName: 'Van Cozy',
      'phone[0].phone': '+33678987654',
      'email[0].email': 'jcvc@cozy.cloud',
      'address[0].address': '18 rue des fleurs, Pecado',
      cozy: 'https://jcvd.cozy.cloud',
      company: 'Cozy CLoud',
      jobTitle: 'Dreamer',
      birthday: '31/12/2015',
      note: 'Whatever.'
    }

    Object.keys(fields).forEach(fieldName => {
      const candidates = form.find(`TextField[name='${fieldName}']`)
      const field = candidates.length === 1 ? candidates : candidates.first()
      field.props().onChange({ target: { value: fields[fieldName] } })
    })

    form.find('form').simulate('submit')
  })

  it('should change inputs value', async () => {
    const testFields = {
      Firstname: 'Jean-Claude',
      Lastname: 'Van Cozy',
      Phone: '+33678987654',
      Email: 'jcvc@cozy.cloud',
      Address: '18 rue des fleurs, Pecado',
      'Cozy URL': 'https://jcvd.cozy.cloud',
      Company: 'Cozy CLoud',
      jobTitle: 'Dreamer',
      Birthday: '31/12/2015',
      Notes: 'Whatever.'
    }

    const props = setup({ contact: {} })
    const jsx = (
      <AppLike client={client}>
        <ContactForm {...props} />
      </AppLike>
    )
    const { findByDisplayValue, getByLabelText } = render(jsx)

    labels.map(label => {
      fireEvent.change(getByLabelText(label), {
        target: { value: testFields[label] }
      })
    })

    expect(await findByDisplayValue(testFields['Firstname'])).toBeTruthy()
    expect(await findByDisplayValue(testFields['Notes'])).toBeTruthy()
  })

  it('should submit empty fields', () => {
    const expected = {
      address: [],
      birthday: undefined,
      company: '',
      cozy: [],
      displayName: '',
      email: [],
      fullname: '',
      indexes: { byFamilyNameGivenNameEmailCozyUrl: '' },
      jobTitle: '',
      metadata: { cozy: true, version: 1 },
      name: { familyName: undefined, givenName: undefined },
      note: '',
      phone: [],
      relationships: { groups: { data: [] } }
    }

    const onSubmit = contact => {
      expect(contact).toEqual(expected)
    }

    const form = mount(
      <I18n lang="en" dictRequire={dictRequire}>
        <ContactForm onSubmit={onSubmit} onCancel={() => {}} />
      </I18n>
    )

    form.find('form').simulate('submit')
  })
})
