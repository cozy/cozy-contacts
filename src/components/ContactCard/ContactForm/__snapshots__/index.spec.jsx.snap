// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactForm should fill form with contact data 1`] = `
<div>
  <form
    className="contact-form"
    onSubmit={[Function]}
  >
    <div
      className="contact-form__fields"
    >
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          Firstname
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="givenName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="text"
              value="John"
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          Lastname
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="familyName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={true}
              type="text"
              value="Doe"
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Phone
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="phone[0].phone"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="tel"
                value="+33 (2)0 90 00 54 04"
              />
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
                data-valid={true}
                name="phone[0].phoneLabel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Label"
                type="text"
                value=""
              />
            </div>
            <button
              className="contact-form__meta-button contact-form__meta-button--remove"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="styles__icon___23x3R"
                height="12"
                style={
                  Object {
                    "fill": "var(--coolGrey)",
                  }
                }
                width="12"
              >
                <use
                  xlinkHref="#cross-small"
                />
              </svg>
            </button>
          </div>
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="phone[1].phone"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="tel"
                value="+33 6 77 11 22 33"
              />
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
                data-valid={true}
                name="phone[1].phoneLabel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Label"
                type="text"
                value=""
              />
            </div>
            <button
              className="contact-form__meta-button contact-form__meta-button--remove"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="styles__icon___23x3R"
                height="12"
                style={
                  Object {
                    "fill": "var(--coolGrey)",
                  }
                }
                width="12"
              >
                <use
                  xlinkHref="#cross-small"
                />
              </svg>
            </button>
          </div>
          <button
            className="styles__c-btn___-2Vnj styles__c-btn--text___2Vp-2"
            onClick={[Function]}
            title={null}
            type="button"
          >
            <span>
              <svg
                aria-hidden={true}
                className="styles__icon___23x3R"
                focusable="false"
                height="16"
                style={Object {}}
                width="16"
              >
                <use
                  xlinkHref="#plus"
                />
              </svg>
              <span
                className={null}
              >
                Add a phone
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Email
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="email[0].email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="email"
                value="john.doe@cozycloud.cc"
              />
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
                data-valid={true}
                name="email[0].emailLabel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Label"
                type="text"
                value=""
              />
            </div>
            <button
              className="contact-form__meta-button contact-form__meta-button--remove"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="styles__icon___23x3R"
                height="12"
                style={
                  Object {
                    "fill": "var(--coolGrey)",
                  }
                }
                width="12"
              >
                <use
                  xlinkHref="#cross-small"
                />
              </svg>
            </button>
          </div>
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="email[1].email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="email"
                value="john.doe@posteo.net"
              />
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
                data-valid={true}
                name="email[1].emailLabel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Label"
                type="text"
                value="personal"
              />
            </div>
            <button
              className="contact-form__meta-button contact-form__meta-button--remove"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="styles__icon___23x3R"
                height="12"
                style={
                  Object {
                    "fill": "var(--coolGrey)",
                  }
                }
                width="12"
              >
                <use
                  xlinkHref="#cross-small"
                />
              </svg>
            </button>
          </div>
          <button
            className="styles__c-btn___-2Vnj styles__c-btn--text___2Vp-2"
            onClick={[Function]}
            title={null}
            type="button"
          >
            <span>
              <svg
                aria-hidden={true}
                className="styles__icon___23x3R"
                focusable="false"
                height="16"
                style={Object {}}
                width="16"
              >
                <use
                  xlinkHref="#plus"
                />
              </svg>
              <span
                className={null}
              >
                Add an email
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Address
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="address[0].address"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="text"
                value="94 Hinton Road 05034 Fresno, Singapore"
              />
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
                data-valid={true}
                name="address[0].addressLabel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Label"
                type="text"
                value="Home"
              />
            </div>
            <button
              className="contact-form__meta-button contact-form__meta-button--remove"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="styles__icon___23x3R"
                height="12"
                style={
                  Object {
                    "fill": "var(--coolGrey)",
                  }
                }
                width="12"
              >
                <use
                  xlinkHref="#cross-small"
                />
              </svg>
            </button>
          </div>
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="address[1].address"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="text"
                value="426 Runolfsson Knolls 84573 Port Easter  Cocos (Keeling) Islands"
              />
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
                data-valid={true}
                name="address[1].addressLabel"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder="Label"
                type="text"
                value="Work"
              />
            </div>
            <button
              className="contact-form__meta-button contact-form__meta-button--remove"
              onClick={[Function]}
              type="button"
            >
              <svg
                className="styles__icon___23x3R"
                height="12"
                style={
                  Object {
                    "fill": "var(--coolGrey)",
                  }
                }
                width="12"
              >
                <use
                  xlinkHref="#cross-small"
                />
              </svg>
            </button>
          </div>
          <button
            className="styles__c-btn___-2Vnj styles__c-btn--text___2Vp-2"
            onClick={[Function]}
            title={null}
            type="button"
          >
            <span>
              <svg
                aria-hidden={true}
                className="styles__icon___23x3R"
                focusable="false"
                height="16"
                style={Object {}}
                width="16"
              >
                <use
                  xlinkHref="#plus"
                />
              </svg>
              <span
                className={null}
              >
                Add an address
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Cozy URL
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="cozy"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="url"
              value="https://johndoe.mycozy.cloud"
            />
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__label-input"
              data-valid={true}
              name="cozyLabel"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder="Label"
              type="text"
              value="MyCozy"
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Company
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="company"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="text"
              value="Cozy cloud"
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Birthday
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="birthday"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="date"
              value="1999-5-1"
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Notes
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <textarea
              className="styles__c-textarea___D7EEH contact-form__input"
              data-valid={true}
              name="note"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="textarea"
              value="Atque cupiditate saepe omnis quos ut molestiae labore voluptates omnis."
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="contact-form__footer"
    >
      <div>
        <button
          className="styles__c-btn___-2Vnj styles__c-btn--secondary___3Br_N"
          onClick={[Function]}
          title={null}
          type="button"
        >
          <span>
            <span
              className={null}
            >
              Cancel
            </span>
          </span>
        </button>
        <button
          className="styles__c-btn___-2Vnj"
          title={null}
          type="submit"
        >
          <span>
            <span
              className={null}
            >
              Save
            </span>
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`ContactForm should match snapshot 1`] = `
<div>
  <form
    className="contact-form"
    onSubmit={[Function]}
  >
    <div
      className="contact-form__fields"
    >
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          Firstname
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="givenName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          Lastname
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="familyName"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={true}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Phone
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="phone[0].phone"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="tel"
                value=""
              />
            </div>
          </div>
          <button
            className="styles__c-btn___-2Vnj styles__c-btn--text___2Vp-2"
            onClick={[Function]}
            title={null}
            type="button"
          >
            <span>
              <svg
                aria-hidden={true}
                className="styles__icon___23x3R"
                focusable="false"
                height="16"
                style={Object {}}
                width="16"
              >
                <use
                  xlinkHref="#plus"
                />
              </svg>
              <span
                className={null}
              >
                Add a phone
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Email
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="email[0].email"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="email"
                value=""
              />
            </div>
          </div>
          <button
            className="styles__c-btn___-2Vnj styles__c-btn--text___2Vp-2"
            onClick={[Function]}
            title={null}
            type="button"
          >
            <span>
              <svg
                aria-hidden={true}
                className="styles__icon___23x3R"
                focusable="false"
                height="16"
                style={Object {}}
                width="16"
              >
                <use
                  xlinkHref="#plus"
                />
              </svg>
              <span
                className={null}
              >
                Add an email
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Address
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__meta-wrapper"
          >
            <div
              className="contact-form__input-wrapper"
            >
              <input
                className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
                data-valid={true}
                name="address[0].address"
                onBlur={[Function]}
                onChange={[Function]}
                onFocus={[Function]}
                placeholder=""
                required={false}
                type="text"
                value=""
              />
            </div>
          </div>
          <button
            className="styles__c-btn___-2Vnj styles__c-btn--text___2Vp-2"
            onClick={[Function]}
            title={null}
            type="button"
          >
            <span>
              <svg
                aria-hidden={true}
                className="styles__icon___23x3R"
                focusable="false"
                height="16"
                style={Object {}}
                width="16"
              >
                <use
                  xlinkHref="#plus"
                />
              </svg>
              <span
                className={null}
              >
                Add an address
              </span>
            </span>
          </button>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Cozy URL
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="cozy"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="url"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Company
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="company"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="text"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Birthday
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <input
              className="styles__c-input-text___3TAv1 styles__c-input-text--large___28EaR contact-form__input"
              data-valid={true}
              name="birthday"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="date"
              value=""
            />
          </div>
        </div>
      </div>
      <div
        className="contact-form__field"
      >
        <label
          className="contact-form__label"
        >
          <svg
            className="contact-form__icon styles__icon___23x3R"
            height="16"
            style={
              Object {
                "fill": "var(--coolGrey)",
              }
            }
            width="16"
          >
            <use
              xlinkHref="#icon-0"
            />
          </svg>
          Notes
        </label>
        <div
          className="contact-form__inputs-wrapper"
        >
          <div
            className="contact-form__input-wrapper"
          >
            <textarea
              className="styles__c-textarea___D7EEH contact-form__input"
              data-valid={true}
              name="note"
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              placeholder=""
              required={false}
              type="textarea"
              value=""
            />
          </div>
        </div>
      </div>
    </div>
    <div
      className="contact-form__footer"
    >
      <div>
        <button
          className="styles__c-btn___-2Vnj styles__c-btn--secondary___3Br_N"
          onClick={[Function]}
          title={null}
          type="button"
        >
          <span>
            <span
              className={null}
            >
              Cancel
            </span>
          </span>
        </button>
        <button
          className="styles__c-btn___-2Vnj"
          title={null}
          type="submit"
        >
          <span>
            <span
              className={null}
            >
              Save
            </span>
          </span>
        </button>
      </div>
    </div>
  </form>
</div>
`;

exports[`ContactForm should submit a well formatted contact 1`] = `
Object {
  "address": Array [
    Object {
      "formattedAddress": "18 rue des fleurs, Pecado",
      "primary": true,
      "type": undefined,
    },
  ],
  "birthday": "31/12/2015",
  "company": "Cozy CLoud",
  "cozy": Array [
    Object {
      "label": undefined,
      "primary": true,
      "url": "https://jcvd.cozy.cloud",
    },
  ],
  "email": Array [
    Object {
      "address": "jcvc@cozy.cloud",
      "primary": true,
      "type": undefined,
    },
  ],
  "fullname": "Jean-Claude Van Cozy",
  "metadata": Object {
    "cozy": true,
    "version": 1,
  },
  "name": Object {
    "familyName": "Van Cozy",
    "givenName": "Jean-Claude",
  },
  "note": "Whatever.",
  "phone": Array [
    Object {
      "number": "+33678987654",
      "primary": true,
      "type": undefined,
    },
  ],
  "relationships": Object {
    "groups": Object {
      "data": Array [],
    },
  },
}
`;

exports[`ContactForm should submit empty fields 1`] = `
Object {
  "address": Array [],
  "birthday": undefined,
  "company": "",
  "cozy": Array [],
  "email": Array [],
  "fullname": "",
  "metadata": Object {
    "cozy": true,
    "version": 1,
  },
  "name": Object {
    "familyName": undefined,
    "givenName": undefined,
  },
  "note": "",
  "phone": Array [],
  "relationships": Object {
    "groups": Object {
      "data": Array [],
    },
  },
}
`;
